### ISSUE TITLE
invalid_intent_id when changing payment method for a 3DSecure one

### ISSUE THREAD

**Issue by agenceKanvas** [woocommerce/woocommerce-gateway-stripe#3624]
**Describe the bug**
I have a customer with an active subscription : 
![Image](https://github.com/user-attachments/assets/a8a0194f-dc0c-4764-97bc-52c432b29c7d)

He tries to change his payment method for a card with 3DS (stripe card number 4000000000003220)
![Image](https://github.com/user-attachments/assets/51378174-f962-40fd-b677-2403a543d052)

The 3DS popup is triggered :
![Image](https://github.com/user-attachments/assets/2e597e36-89b3-497f-9aa8-f76dc2b07fed)

But in the background (JS console), I have an error : 
![Image](https://github.com/user-attachments/assets/c895d8db-fafa-4a14-9e01-fceaae58efe2)

The result is that when I validate the 3DS, the info is sent to Stripe with a success, but as there is already an error on the page, the switch on WooCommerce fails, I have a payment intent error display, and the subscription is set on pending. 
The error displayed in the notice is "invalid_intent_id" 

The same thing is happening on the production version 

Last version of WC, WC Stripe, Subscriptions and membership used.

Do you have any idea ? 

thanks a lot !

**Comment by jvinzent** [comment]
(for us it looks like #3312, included in #3313, we are confused why this occurs @james-allan )

**Comment by agenceKanvas** [comment]
Can't figure out what's the problem ... If you have some pointers, I'm all ears :)

**Comment by agenceKanvas** [comment]
Hi ! 
Am I the only one getting this error ? Any idea on how I could give you more infos to pinpoint the problem ? 
I still have the same error with the last version of every plugins

**Comment by agenceKanvas** [comment]
I've, of course, tried to disable all my plugins and use the 2024 theme. Still the same problem. 
Maybe some differences on the Stripe configuration, but I will need your help to debug that. Thanks a lot !

**Comment by hsingyuc** [comment]
@agenceKanvas & @jvinzent, thank you for reporting this issue and sharing the information. I can reproduce the issue and will be working on it.

**Comment by hsingyuc** [comment]
@agenceKanvas, do you have WooCommerce Payments enabled? Are you able to reproduce the issue with WooCommerce Payments disabled?

**Comment by agenceKanvas** [comment]
We don't use WooCommerce Payments @hsingyuc . Only Stripe

**Comment by hsingyuc** [comment]
I can only reproduce the issue if I checkout with WooPayments and change the payment method to a 3DS one. I can not reproduce the issue without WooPayments.

**Comment by agenceKanvas** [comment]
@hsingyuc It worked on your side when you change a payment method to a 3DS required ?

**Comment by agenceKanvas** [comment]
To give you more info to compare, this is what I have on Stripe :
![Image](https://github.com/user-attachments/assets/2312d318-04f0-4d66-9c5a-0a8955dcef2c)

I've looked into all of the webhooks and I can't find anywhere the intent_id

In which one should it be ? 

Thanks !!

**Comment by hsingyuc** [comment]
>@hsingyuc It worked on your side when you change a payment method to a 3DS required ?

Correct.

Here's what I tried to reproduce the error, let me know if I missed something here. Thanks!
1. WC 9.5.2, WC memberships 1.26.9, Stripe 9.1.1, and Subscriptions 7.1.0
2. As a shopper, purchase a Subscriptions product with Stripe, card number 4242424242424242 on classic checkout page
3. As a shopper, log in on the merchant site and find the Subscription order I just made under My Subscription.
4. Click on the Change payment button
5. Use the 3DS card 4000000000003220
6. See payment method updated

<img width="663" alt="Image" src="https://github.com/user-attachments/assets/86deead0-211f-4a54-a243-21cc9fd4902f" />

**Comment by agenceKanvas** [comment]
Thanks for your reply @hsingyuc . 
Using the same card number I have a invalid_intent_id. 

This is the error I have in my order notes (translated from french) :

A payment with the ID seti_1QjGjrCV9djTjhnHb80JbqvR was used in an attempt to pay for this order. This payment intent ID does not correspond to any payment for this order, so it has been ignored and the order has not been updated.

Shouldn't I have a pi_ instead of a seti_ ? Where should I look please ? We're stuck since a long time with that :/

**Comment by malithsen** [comment]
I've also not been able to replicate this issue so far. 

> Shouldn't I have a pi_ instead of a seti_ ? Where should I look please ?

@agenceKanvas After you click the "Complete" button in the 3DS popup, there should be a `wc_stripe_confirm_change_payment` AJAX request, containing the `order_id` for your order and the `intent_id`  in the form `seti_*`. 

![Image](https://github.com/user-attachments/assets/16f81df5-8b61-4a3e-9d15-b8d01e8c2a45)

The plugin handles this request in [this function](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/a92867d7bbc36290728fc25c3a8da52b52320e3a/includes/class-wc-stripe-intent-controller.php#L1085) to create the payment method token from the setup intent. 

Can you check if you have the above AJAX request via your browser dev tools? If so, can you also check if [confirm_change_payment_from_setup_intent_ajax](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/a92867d7bbc36290728fc25c3a8da52b52320e3a/includes/class-wc-stripe-intent-controller.php#L1085) processes this request.

**Comment by agenceKanvas** [comment]
Thanks @malithsen ! 
No, I don't have that request poping : 
![Image](https://github.com/user-attachments/assets/75b4708a-b593-4148-b030-ab2ae4c701ab)

And I've added some logs in the confirm_change_payment_from_setup_intent and it never fires.

**Comment by malithsen** [comment]
@agenceKanvas Thanks for checking. That narrows it down a little bit!

So it seems like, for some reason, the frontend is making an `update_order_status` request when it should make a `confirm_change_payment` request for you. The switch happens [here](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/a92867d7bbc36290728fc25c3a8da52b52320e3a/client/api/index.js#L337). The request is determined based on the presence of `order-pay` substring in the URL and the presence of an element with the id `#wc-stripe-change-payment-method`.

Can you inspect the "Change payment method" button and see if you have two hidden input fields like below
- input element with id `#wc-stripe-change-payment-method`
- input element with value set to a partial URL. This URL should contain the text `order-pay`. 

![Image](https://github.com/user-attachments/assets/3503e50d-7ced-4c38-8bf5-ba6802604a32)

**Comment by agenceKanvas** [comment]
@malithsen  Like that ? 

![Image](https://github.com/user-attachments/assets/ec703e0b-9aba-4d54-895e-b05ce424562b)

**Comment by agenceKanvas** [comment]
The endpoint /abonnement/commande/paiement is the right one

**Comment by malithsen** [comment]
@agenceKanvas I can replicate the issue now. Just one last check:
In WooCommerce -> Settings -> Advanced -> Checkout endpoints, what do you have for the "Pay" endpoint?

![Image](https://github.com/user-attachments/assets/8a81502f-048d-4c66-a6fe-8e8f2d5d349f)

**Comment by agenceKanvas** [comment]
I have that @malithsen : 

<img width="662" alt="Image" src="https://github.com/user-attachments/assets/8c71b2af-1b9a-4ba6-b698-c91536c29c12" />

**Comment by agenceKanvas** [comment]
@malithsen : Do you think it's because I use a different endpoint ?

**Comment by malithsen** [comment]
@agenceKanvas Yes, there's a hard-coded check that looks for `order-pay` endpoint [here](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/a92867d7bbc36290728fc25c3a8da52b52320e3a/client/api/index.js#L312). I will open a PR to make it respect custom endpoints. But for now, if you update your "Pay" endpoint ("Payer") to `order-pay`, it should fix the error you're getting.

**Comment by agenceKanvas** [comment]
@malithsen Awesome. Thanks a lot ! 
If you need testers with different setup, count me in whenever you need it :)

**Comment by malithsen** [comment]
@agenceKanvas Thanks for helping us figure out the underlying issue. PR #3753 is merged and expected to go out in the next release.
